server/src/main/java/org/elasticsearch/common/MemoryAccountingBytesRefCounted.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common;
+
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.core.AbstractRefCounted;
+
+/**
+ * A ref counted object that accounts for memory usage in bytes and releases the
+ * accounted memory from the circuit breaker when the reference count reaches zero.
+ */
+public final class MemoryAccountingBytesRefCounted extends AbstractRefCounted {
```

Comment:
Bit of a general question here on the safety of this:

We have a mutable `int` field that is not thread safe, but a thread-safe ref-count.
Can we add some comments on how we know that we have a proper barrier/ordering/happens-before-edge/etc. between the `account` calls and the eventual `closeInternal`?

Reply:
Oh it's not safe like this:) Changed it to an AtomicInteger - we write it and read it once every buffer size, but from different threads.

Reply:
Wait but aren't all updates to the value confined to a single thread? All we need is the final update to show up at the point of closing?
+ weren't we just worried about the size of this thing, `AtomicInteger` makes it quite a bit heavier?

Reply:
>All we need is the final update to show up at the point of closing?
Yes, this is what we need. Oh, volatile int would be cheaper here! 

>weren't we just worried about the size of this thing, AtomicInteger makes it quite a bit heavier?
Yes. but we're doing it once/batch now (as opposed to for every hit)

Reply:
What do you think if we change `account` to 
```
public void setBytesAndAccount(int bytes, String label) {
        breaker.addEstimateBytesAndMaybeBreak(bytes, label);
        this.bytes = bytes;
}
```
and have the instance `this.bytes` a `volatile int` ?

Reply:
Sketched it here https://github.com/elastic/elasticsearch/pull/121920/commits/f366e7f0d7729a21bed0f680f185b533b1d30b35